API: Fix  "Updated by" Column in dashboard versions table

papagian opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

7. If your PR content should be added to the What's New document for the next major or minor release, add the **add to what's new** label to your PR. Note that you should add this label to the main PR that introduces the feature; do not add this label to smaller PRs for the feature.

-->

**What is this feature?**

This fix modifies the `GET /dashboards/uid/{uid}/versions/` endpoint and returns the user login (string) in the `created_by` JSON field - it used to return the user sequential ID (int).
More specifically it changes the returned response to `[]dashver.DashboardVersionMeta` (similarly to the `GET /dashboards/uid/{uid}/versions/{DashboardVersionID}` endpoint)
Also, it caches the result for 5 seconds.

In addition to this, it updates (recreates) the OpenAPI specifications. Note that the specifications contain additional changes for other modifications introduced in the API outside the scope of this but the specifications weren't updated as well.

**Why do we need this feature?**

It populates "Updated by" column in dashboard version table with the user login instead of the user sequential ID

**Who is this feature for?**

[Add information on what kind of user the feature is for.]

**Which issue(s) does this PR fix?**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #60935

**Special notes for your reviewer:**
The implementation slightly differs from `GET /dashboards/uid/{uid}/versions/{DashboardVersionID}` endpoint:
https://github.com/grafana/grafana/blob/54d7e95be19d5f0cbd3c7d697966b27bd8feac20/pkg/api/dashboard.go#L753-L756
that calls the `getUserLogin ()` which in case of error returns the `Anonymous` (regardless of the error) that does not seem right
https://github.com/grafana/grafana/blob/54d7e95be19d5f0cbd3c7d697966b27bd8feac20/pkg/api/dashboard.go#L261-L268
**Updated** 2023-03-29: I have changed my implementation to align with `GET /dashboards/uid/{uid}/versions/{DashboardVersionID}` implementation (we re-consider in a follow up whether defaulting to `Anonymous` is the right option)

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.
- [ ] There are no known compatibility issues with older supported versions of Grafana, or plugins.
- [ ] It passes the [Hosted Grafana feature readiness review](https://docs.google.com/document/d/1QL9Ly8KnXzpb6ISbg49pTODRO5mhA5tkkfIZVX6pqQU/edit) for observability, scalability, performance, and security.

papagian committed:
API: Fix dashboard versions created by field

papagian committed:
Add test

papagian committed:
Merge remote-tracking branch 'origin/main' into papagian/fix-dashboard-versions-created-by

papagian committed:
Add test

papagian committed:
Fixup

papagian committed:
Fix test

Remove previously introduced change that breaks the test
because of the way the http server is initialized for the test:
missing cache service

suntala commented on pkg/api/dashboard_test.go:
```diff
@@ -935,6 +943,129 @@ func TestDashboardAPIEndpoint(t *testing.T) {
+			var versions []*dashver.DashboardVersionDTO
+			err := json.NewDecoder(sc.resp.Body).Decode(&versions)
+			require.NoError(t, err)
+			for _, v := range versions {
```
What do you think about checking the other fields of `v` somehow? Or even just checking that `len(versions) == 2`? I didn't see any checks along these lines in this test file.

Same comment applies to the other two test cases.

papagian commented on pkg/api/dashboard.go:
```diff
@@ -665,6 +666,7 @@ func (hs *HTTPServer) GetDashboardVersions(c *contextmodel.ReqContext) response.
 
+	const cacheKey = "userLogin-"
 	for _, version := range res {
 		if version.RestoredFrom == version.Version {
```
good catch!

papagian committed:
Fix implementation and tests after code review

Change response to return `[]DashboardVersionMeta` instead of `[]*DashboardVersionDTO`
and revert previously introduced
change to `DashboardVersionDTO` json tags.

papagian committed:
Update OpenAPI specs

sakjur reviewed (approved):
LGTM, left a comment with a suggestion for making this less likely to trigger a panic.

papagian commented on pkg/api/dashboard.go:
```diff
@@ -660,25 +661,52 @@ func (hs *HTTPServer) GetDashboardVersions(c *contextmodel.ReqContext) response.
+		creator := anonString
+		if version.CreatedBy > 0 {
+			key := fmt.Sprintf("%s%d", cacheKey, version.CreatedBy)
+			if cached, found := hs.CacheService.Get(key); found {
+				creator = cached.(string)
+			} else {
+				creator = hs.getUserLogin(c.Req.Context(), version.CreatedBy)
+				if creator != anonString {
+					hs.CacheService.Set(key, creator, time.Second*5)
+				}
+			}
```
agree with most of them except:
> I also think we should cache even if the creator is "anonymous".

because the function returns `Anonymous` for any error case:
https://github.com/grafana/grafana/blob/54d7e95be19d5f0cbd3c7d697966b27bd8feac20/pkg/api/dashboard.go#L261-L268

sakjur replied:
:'( gotcha, we should probably change _that_ at some point, but good catch

papagian committed:
Apply suggestion from code review
